ast2vec/repo2/coocc.md

```diff
@@ -0,0 +1,45 @@
+# Co-occurrence matrix calculation.
+This document covers an algorithm to calculate identifiers co-occurrence matrix based on UAST from bblfsh.
+It's a preparation step to calculate embeddings for them.
+
+There are several files related to this process:
+* ast2vec/repo2/cooccbase.py - base class with main logic for UAST traversing
+* ast2vec/repo2/coocc.py - construct matrix from scratch
+* ast2vec/repo2/voccoocc.py - construct matrix with provided vocabulary
+* ast2vec/voccoocc.py - model for modelforge (way to save/load model)
+* ast2vec/coocc.py - model for modelforge
+
+Let's explain them one by one.
+## ast2vec/repo2/cooccbase.py
+It contains logic for UAST traversing.
+The base algorithm to compute co-occurrence matrix is based on BFS.
+Our goal is create context for each identifier.
+
+![initial UAST](../../doc/co-occ_expl1.png)
+
+As we can see at image above - we have some nodes without identifiers that we have to filter out for this task.
+After we filter them out we should have situation like this:
+
+![after filtering](../../doc/co-occ_expl2.png)
+
+And then we just connect root with children and all children at the same level.
```

Comment:
We connect the root with its children then, and all the children appear on the same level